### PR TITLE
add post

### PR DESCRIPTION
**Pull request by skeydan**

### CODE REVIEW

File: _posts/2022-10-06-audio-classification-torch/audio-classification.Rmd [hunk 1]
```diff
@@ -0,0 +1,472 @@
+---
+title: "Audio classification with torch"
+description: >
+  Learn how to classify speech utterances with torch, making use of domain knowledge and deep learning. This post is a condensed version of the same-topic chapter in the book to be published with CRC Press, Deep Learning and Scientific Computing with R torch.
```
Comment (batpigandme) [inline _posts/2022-10-06-audio-classification-torch/audio-classification.Rmd:4]:
```suggestion
  Learn how to classify speech utterances with torch, making use of domain knowledge and deep learning. This post is a condensed version of the corresponding chapter in the forthcoming book, Deep Learning and Scientific Computing with R torch, to be published by CRC Press. 
```

Comment (batpigandme) [reply _posts/2022-10-06-audio-classification-torch/audio-classification.Rmd:4]: Tried to re-jig the words to frontload the most important info (like the name of the book, versus the name of the publisher).

File: _posts/2022-10-06-audio-classification-torch/audio-classification.Rmd [hunk 2]
```diff
@@ -0,0 +1,472 @@
...
+description: >
+  Learn how to classify speech utterances with torch, making use of domain knowledge and deep learning. This post is a condensed version of the same-topic chapter in the book to be published with CRC Press, Deep Learning and Scientific Computing with R torch. 
+author:
+  - name: Sigrid Keydana
+    affiliation: RStudio
+    affiliation_url: https://www.rstudio.com/
+slug: keydanatorchaudio
+date: 2022-10-06
+categories:
+  - Torch
```
Comment (batpigandme) [inline _posts/2022-10-06-audio-classification-torch/audio-classification.Rmd:12]: I haven't checked, but is Torch always capitalized in the tags? I'm inferring that it's capitalized when referring to Torch (the library) versus small-t torch the R package (bonus confusion for the fact that the native package is also small-t it looks like from [the Wikipedia page](https://en.wikipedia.org/wiki/Torch_(machine_learning))).

I think it's fine either way—it's obvious from the fact that both Torch and R are in the tags that you're going to discuss the R package for it—just checking for consistency.

File: _posts/2022-10-06-audio-classification-torch/audio-classification.Rmd [hunk 3]
```diff
@@ -0,0 +1,472 @@
...
+preview: images/squirrel.jpg
+---
+
+```{r setup, include=FALSE}
+knitr::opts_chunk$set(echo = TRUE, eval = FALSE, fig.width = 6, fig.height = 6)
+```
+
+## Variations on a theme
+
+[Simple Audio Classification with Keras](https://blogs.rstudio.com/ai/posts/2018-06-06-simple-audio-classification-keras/), [Audio classification with Keras: Looking closer at the non-deep learning parts](https://blogs.rstudio.com/ai/posts/2019-02-07-audio-background/), [Simple audio classification with torch](https://blogs.rstudio.com/ai/posts/2021-02-04-simple-audio-classification-with-torch/): No, this is not the first post, on this blog, that introduces speech classification using deep learning. With two of those posts (the "applied" ones) it shares the general setup, the type of deep-learning architecture employed, and the dataset used. With the third, it has in common the interest in the ideas and concepts involved. Each of these posts has a different focus, -- should you read this one?
```
Comment (batpigandme) [inline _posts/2022-10-06-audio-classification-torch/audio-classification.Rmd:29]: It looks like the three of you actually do use different capitalization rules for your post titles. But, since you and Athos are in the majority, I think Daniel can get overriden here for consistency's sake.

Comment (batpigandme) [inline _posts/2022-10-06-audio-classification-torch/audio-classification.Rmd:29]:
```suggestion
[Simple audio classification with Keras](https://blogs.rstudio.com/ai/posts/2018-06-06-simple-audio-classification-keras/), [Audio classification with Keras: Looking closer at the non-deep learning parts](https://blogs.rstudio.com/ai/posts/2019-02-07-audio-background/), [Simple audio classification with torch](https://blogs.rstudio.com/ai/posts/2021-02-04-simple-audio-classification-with-torch/): No, this is not the first post on this blog that introduces speech classification using deep learning. With two of those posts (the "applied" ones) it shares the general setup, the type of deep-learning architecture employed, and the dataset used. With the third, it has in common the interest in the ideas and concepts involved. Each of these posts has a different focus -- should you read this one?
```

File: _posts/2022-10-06-audio-classification-torch/audio-classification.Rmd [hunk 4]
```diff
@@ -0,0 +1,472 @@
...
+
+Well, of course I can't say "no" here -- all the more that here, you have an abbreviated and condensed version of the same-topic chapter that is part of the book to be published by CRC Press, *Deep Learning and Scientific Computing with R `torch`*. By way of comparison with the previous post that used `torch`, written by the creator and maintainer of `torchaudio`, Athos Damiani, significant developments have taken place in the `torch` ecosystem, the end result being that the code got a lot easier (especially in the model training part). That said, let's end the preamble already, and plunge into the topic!
```
Comment (batpigandme) [inline _posts/2022-10-06-audio-classification-torch/audio-classification.Rmd:31]:
```suggestion
Well, of course I can't say "no" -- all the more so because, here, you have an abbreviated and condensed version of the chapter on this topic in the forthcoming book from CRC Press, *Deep Learning and Scientific Computing with R `torch`*. By way of comparison with the previous post that used `torch`, written by the creator and maintainer of `torchaudio`, Athos Damiani, significant developments have taken place in the `torch` ecosystem, the end result being that the code got a lot easier (especially in the model training part). That said, let's end the preamble already, and plunge into the topic!
```

Comment (batpigandme) [reply _posts/2022-10-06-audio-classification-torch/audio-classification.Rmd:31]: Tried to remove the double use of "here", and re-work things to avoid "same-topic chapter" (which sounds like a strange phrase to my ear).

File: _posts/2022-10-06-audio-classification-torch/audio-classification.Rmd [hunk 5]
```diff
@@ -0,0 +1,472 @@
...
+
+## Inspecting the data
+
+We use the *speech commands* dataset (@abs-1804-03209) that comes with [`torchaudio`](https://github.com/curso-r/torchaudio). The dataset holds recordings of thirty different, one- or two-syllable words, uttered by different speakers; there are about 65,000 audio files overall. Our task will be to predict, from the audio solely, which of thirty possible words was pronounced.
```
Comment (batpigandme) [inline _posts/2022-10-06-audio-classification-torch/audio-classification.Rmd:35]:
```suggestion
We use the *speech commands* dataset (@abs-1804-03209) that comes with [`torchaudio`](https://github.com/curso-r/torchaudio). The dataset holds recordings of thirty different one- or two-syllable words, uttered by different speakers. There are about 65,000 audio files overall. Our task will be to predict, from the audio solely, which of thirty possible words was pronounced.
```

Comment (batpigandme) [reply _posts/2022-10-06-audio-classification-torch/audio-classification.Rmd:35]: You know me, always out to destroy semi-colons 😈😉

File: _posts/2022-10-06-audio-classification-torch/audio-classification.Rmd [hunk 6]
```diff
@@ -0,0 +1,472 @@
...
+    )
+  ) +
+  xlab("time") +
+  ylab("amplitude") +
+  theme_minimal()
+```
+
+![The spoken word "bird", in time-domain representation.](images/audio-bird-waveform.png){#fig-audio-bird-waveform fig-alt="A sound wave, displaying amplitude over time."}
+
+What we see is a sequence of amplitudes, reflecting the sound wave produced by someone saying "bird". Put differently, we have here a time series of "loudness values". Even for experts, guessing *which* word resulted in those amplitudes is an impossible task. This is where domain knowledge comes in. The expert may not be able to make much of the signal *in this representation*; but they may know a way to more meaningfully represent it.
```
Comment (batpigandme) [inline _posts/2022-10-06-audio-classification-torch/audio-classification.Rmd:118]: Really great paragraph!

File: _posts/2022-10-06-audio-classification-torch/audio-classification.Rmd [hunk 7]
```diff
@@ -0,0 +1,472 @@
...
+
+## Two equivalent representations
+
+Imagine that instead of as a sequence of amplitudes over time, the above wave were represented in a way that had no information about time at all. Next, imagine we took that representation and tried to recover the original signal. For that to be possible, the new representation would somehow have to contain "just as much" information as the wave we started from. That "just as much" is obtained from the *Fourier Transform*, and it consists of the magnitudes and phase shifts of the different *frequencies* that make up the signal.
```
Comment (batpigandme) [inline _posts/2022-10-06-audio-classification-torch/audio-classification.Rmd:122]: > the above wave were represented

I'm not sure if you want plural ("waves were") or singular ("wave was"), but the noun and verb should agree, either way.

Otherwise, really nice explanation.

File: _posts/2022-10-06-audio-classification-torch/audio-classification.Rmd [hunk 8]
```diff
@@ -0,0 +1,472 @@
...
+    )
+  ) +
+  xlab("frequency") +
+  ylab("magnitude") +
+  theme_minimal()
+```
+
+![The spoken word "bird", in frequency-domain representation.](images/audio-bird-dft.png){#fig-audio-bird-dft fig-alt="A curve showing magnitude for Fourier bins from 0 to 8000. Above 4000, nearly all are zero."}
+
+From this, alternate, representation, we could go back to the original sound wave by taking the frequencies present in the signal, weighting them according to their coefficients, and adding them up. But in sound classification, timing information must surely matter; we don't really want to throw it away.
```
Comment (batpigandme) [inline _posts/2022-10-06-audio-classification-torch/audio-classification.Rmd:159]:
```suggestion
From this alternate representation, we could go back to the original sound wave by taking the frequencies present in the signal, weighting them according to their coefficients, and adding them up. But in sound classification, timing information must surely matter; we don't really want to throw it away.
```

File: _posts/2022-10-06-audio-classification-torch/audio-classification.Rmd [hunk 9]
```diff
@@ -0,0 +1,472 @@
...
+
+In fact, what really would help us is a synthesis of both representations; some sort of "have your cake and eat it, too". What if we could divide the signal into small chunks, and run the Fourier Transform on each of them? As you may have guessed from this lead-up, this indeed is something we can do; and the representation it creates is called the *spectrogram*.
+
+With a spectrogram, we still keep some time-domain information -- some, since there is an unavoidable loss in granularity. On the other hand, for each of the time segments, we learn about their spectral composition. There's an important point to be made, though. The resolutions we get in *time* versus in *frequency*, respectively, are inversely related. If we split up the signals into many chunks (called "windows"), the frequency representation per window will not be very fine-grained. Conversely, if we want to get better resolution in the frequency domain, we have to choose longer windows, thus losing information about how spectral composition varies over time. What sounds like a big problem -- and in many cases, will be -- won't be one for us, though, as you'll see very soon.
```
Comment (batpigandme) [inline _posts/2022-10-06-audio-classification-torch/audio-classification.Rmd:165]: 👍

File: _posts/2022-10-06-audio-classification-torch/audio-classification.Rmd [hunk 10]
```diff
@@ -0,0 +1,472 @@
...
+main <- paste0("Spectrogram, window size = ", window_size)
+sub <- "Magnitude (square root)"
+mtext(side = 3, line = 2, at = 0, adj = 0, cex = 1.3, main)
+mtext(side = 3, line = 1, at = 0, adj = 0, cex = 1, sub)
+
+```
+
+![The spoken word "bird": Spectrogram.](images/audio-spectrogram.png){#fig-audio-spectrogram fig-alt="A three-dimensional plot where the x-axis represents time, the y-axis, frequency, and color, magnitude."}
+
+We know that we've lost some resolution, in both time and frequency. By displaying the square root of the coefficients' magnitudes, though -- and thus, enhancing sensitivity -- we were still able to obtain a reasonable result. (With the `viridis` color scheme, long-wave shades indicate higher-valued coefficients; short-wave ones, the opposite.)
```
Comment (batpigandme) [inline _posts/2022-10-06-audio-classification-torch/audio-classification.Rmd:215]:
```suggestion
We know that we've lost some resolution in both time and frequency. By displaying the square root of the coefficients' magnitudes, though -- and thus, enhancing sensitivity -- we were still able to obtain a reasonable result. (With the `viridis` color scheme, long-wave shades indicate higher-valued coefficients; short-wave ones, the opposite.)
```

File: _posts/2022-10-06-audio-classification-torch/audio-classification.Rmd [hunk 11]
```diff
@@ -0,0 +1,472 @@
...
+
+Display-wise, of course, the full complex representation is inconvenient; the spectrogram plot would need an additional dimension. But we may well wonder whether a neural network could profit from the additional information contained in the "whole" complex number. After all, when reducing to magnitudes we lose the phase shifts for the individual coefficients, which might contain usable information. In fact, my tests showed that it did; use of the complex values resulted in enhanced classification accuracy. [^1]
+
+[^1]: In the book, I'm comparing three different setups (the third being use of Mel spectrograms). Using complex coefficients for network input was the winner.
```
Comment (batpigandme) [inline _posts/2022-10-06-audio-classification-torch/audio-classification.Rmd:277]:
```suggestion
[^1]: In the book, I compare three different setups (the third being use of Mel spectrograms). Using complex coefficients for network input was the winner.
```

File: _posts/2022-10-06-audio-classification-torch/audio-classification.Rmd [hunk 12]
```diff
@@ -0,0 +1,472 @@
...
+  download = TRUE,
+  power = NULL
+)
+
+dim(ds[1]$x)
+```
+
+    [1]   2 257 101
+
+We have two hundred fifty-seven coefficients for a hundred and one windows; and each coefficient is represented by both its real and imaginary parts.
```
Comment (batpigandme) [inline _posts/2022-10-06-audio-classification-torch/audio-classification.Rmd:294]:
```suggestion
We have 257 coefficients for 101 windows; and each coefficient is represented by both its real and imaginary parts.
```

Comment (batpigandme) [reply _posts/2022-10-06-audio-classification-torch/audio-classification.Rmd:294]: Numbers over ten (or twenty, depending on whom you ask) are written in Arabic numerals. ✅

File: _posts/2022-10-06-audio-classification-torch/audio-classification.Rmd [hunk 13]
```diff
@@ -0,0 +1,472 @@
...
+We can confirm this on the test set:
+
+```{r}
+evaluate(fitted, test_dl)
+```
+
+    loss: 0.2373
+    acc: 0.9324
+
+An interesting question is which words get confused most often. (Of course, even more interesting is how error probabilities are related to features of the spectrograms \-- but this, we have to leave to the *true* domain experts. A nice way of displaying the confusion matrix is to create an alluvial plot. We see the predictions, on the left, "flow into" the target slots. (Target-prediction pairs less frequent than a thousandth of test set cardinality are hidden.)
```
Comment (batpigandme) [inline _posts/2022-10-06-audio-classification-torch/audio-classification.Rmd:464]: > the spectrograms \-- but this

Is that slash meant to be there?

**Review by batpigandme — APPROVED** [review]
Well done! Very accessible reading, even for someone like me!

File: _posts/2022-10-06-audio-classification-torch/audio-classification.Rmd (hunk 1 above)
Comment (skeydan) [reply _posts/2022-10-06-audio-classification-torch/audio-classification.Rmd:4]: thanks! taking it as a compliment for my politeness ;-))))

File: _posts/2022-10-06-audio-classification-torch/audio-classification.Rmd (hunk 2 above)
Comment (skeydan) [reply _posts/2022-10-06-audio-classification-torch/audio-classification.Rmd:12]: it's just the category really that is capitalized (for aesthetical reasons, because it would look weird otherwise)

File: _posts/2022-10-06-audio-classification-torch/audio-classification.Rmd (hunk 3 above)
Comment (skeydan) [reply _posts/2022-10-06-audio-classification-torch/audio-classification.Rmd]: :-))

File: _posts/2022-10-06-audio-classification-torch/audio-classification.Rmd (hunk 4 above)
Comment (skeydan) [reply _posts/2022-10-06-audio-classification-torch/audio-classification.Rmd:31]: > all the more so because,
oh right, that was what I wanted to say, but I didn't get it right :-)

> from CRC Press
prepositions are hard, in English (well, in any language, I guess)

File: _posts/2022-10-06-audio-classification-torch/audio-classification.Rmd (hunk 12 above)
Comment (skeydan) [reply _posts/2022-10-06-audio-classification-torch/audio-classification.Rmd:294]: oh, see, so I over-generalized from some correction you made, in the past :-)

File: _posts/2022-10-06-audio-classification-torch/audio-classification.Rmd (hunk 5 above)
Comment (skeydan) [reply _posts/2022-10-06-audio-classification-torch/audio-classification.Rmd:35]: hehe :-)
Of course, I have mixed thoughts on this ;-)))
Mixed, because in the book, I've taken care to keep sentences short (or at least, that was the intention)

File: _posts/2022-10-06-audio-classification-torch/audio-classification.Rmd (hunk 6 above)
Comment (skeydan) [reply _posts/2022-10-06-audio-classification-torch/audio-classification.Rmd:118]: thanks!!

File: _posts/2022-10-06-audio-classification-torch/audio-classification.Rmd (hunk 7 above)
Comment (skeydan) [reply _posts/2022-10-06-audio-classification-torch/audio-classification.Rmd:122]: it's supposed to be conjunctive (I like the _were_ ;-))

**Comment by skeydan** [comment]
Thank you so much Mara!! And I'm very happy to hear you liked it!
